Problem (fix #294): no spec for subscriptions module (#323)
crypto-com/chain-main: x/subscription/spec/01_concepts.md

<!--
order: 1
-->

# Concepts

## Subscription

> - Merchant create/stop plans on chain, the price is defined on specified token denomination, users should do nessessary token conversion in other ways, e.g. exchange.

Reviewer: just nit -- perhaps for the consistency, "plan owners" instead of merchants (e.g. this could potentially be used for regular donations).
```suggestion
- Subscription plan owners create/stop plans on Chain, the plan prices are defined for specified token denominations, subscribers should do necessary token conversions in other ways, e.g. exchange.
```

> - User subscribe/unsubscribe to plan.

Reviewer: ```suggestion
- Users subscribe/unsubscribe to plan.
```

> - Payments are collected automatically at the start of specified interval, the results of collection are written into block events.

Reviewer: ```suggestion
- Payments are collected automatically at the start of their specified intervals, the results of collections are written into block events.
```

> - Create plan transaction will consume some gases to cover the computational cost of future automatic payment collections.

Reviewer: ```suggestion
- Create plan transaction will consume a gas fee to cover the computational cost of future automatic payment collections.
```

> - In case of automatic collection fails on some subscribers, the chain won't automatic retry later, merchant can either configure it to automatic cancel these subscribers, or retry collection later manually.

Reviewer: ```suggestion
- If the automatic collection mechanism fails on some subscribers, Chain won't automatically retry later, the corresponding plan owner can either configure their plan to automatically cancel these subscribers or manually retry failed collections later.
```

### Intervals

We support specify following types of intervals for plan:

- Fixed seconds interval with offset, like "the beginning of each day offset by UTC+8".

> - [TODO] Calender interval, like "first day of each month".

Reviewer: I'm thinking how this can work with future extensibility -- I guess instead of hardcoding integer, or perhaps have it in protobuf as:
```
message Interval {
  oneof value {
    FixedInterval               fixed                 = 1;
    CalendarInterval              calendar                = 2;
  }
}
```

but perhaps to leave it as it is now -- in the future when calendar intervals are added, one can just use the upgrade module and have the data migration that will convert the stored plans

Author: sorry I didn't noticed these comments, they are collapsed by default.
I'll see if I can manage to bring in some calendar support.

Author: Support crontab syntax directly, that provides best UX.

<!--
order: 1
-->

# Concepts

## Subscription

```go
type Plan struct {
  id int,  // unique identifier
  owner Address,

>   price Coin,

Reviewer: is it a fixed price or a max price?
for a floating exchange rate, one may want to be able to charge a different price each time

Author: merchant want to anchor to a fixed fiat currency price?

Reviewer: yep, that's a very common use case

Author: maybe introduce some oracle?

Reviewer: merchants generally won't do the collection on their own if they receive fiat and it's via a payment provider -- the payment provider does the collection and sets the rate

Author: but if the payment provider is malicious, he can increase the price by change the rate without subscriber's agreement?

Reviewer: generally, that's set in the subscription terms and conditions -- if that happens, the subscriber can cancel subscription and if it broke the terms and conditions, they can go to a small claims court etc.

Author: done, payment provider or merchant can maintain the conversion rate.

Author: changed to fixed denomination.

  interval int, // number of days
}

> type Subscription struct {

Reviewer: How is "expiration" going to be managed -- can merchant / payment gateway cancel/unsubscribe without cancelling the entire plan?

Author: merchant cancel some specific subscriber? it can also choose to stop the off-chain service if it can't collect payments from the subscriber, and resume the service as soon as payment resumes.

Reviewer: > merchant cancel some specific subscriber?
yes

> it can also choose to stop the off-chain service if it can't collect payments from the subscriber, and resume the service as soon as payment resumes.

that doesn't solve the problem, as it'll keep charging (the collection is per plan) / merchant (or payment gateway on its behalf) can't cancel individual subscriptions

Author: both merchant and subscriber can cancel subscription now

Author: both plan and subscription has expiration time now.
